File: rs/src/day23.rs
use std::iter;

fn parse<'a, I, S>(lines: I) -> Option<Vec<u32>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    Some(
        lines
            .into_iter()
            .next()?
            .as_ref()
            .chars()
            .filter_map(|c| c.to_digit(10))
            .collect(),
    )
}

fn step<const N: u32>(arr: &mut [u32], x: u32) -> Option<u32> {
    let a = arr[x as usize];
    let b = arr[a as usize];
    let c = arr[b as usize];
    let y = arr[c as usize];
    let t = iter::successors(Some(x), |t| Some(t.checked_sub(1).unwrap_or(N - 1)))
        .skip(1)
        .find(|t| *t != a && *t != b && *t != c)?;
    let u = *arr.get(t as usize)?;
    arr[x as usize] = y;
    arr[t as usize] = a;
    arr[c as usize] = u;
    Some(y)
}

pub fn part1<'a, I, S>(lines: I) -> Option<u32>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let nums = parse(lines)?;
    let mut arr = [0; 9];
    for (i, x) in nums.iter().enumerate() {
        arr[x.checked_sub(1)? as usize] = *nums.get(i + 1).or_else(|| nums.first())? - 1;
    }
    let mut x = nums[0] - 1;
    for _ in 0..100 {
        x = step::<9>(&mut arr, x)?;
    }
    Some(
        iter::successors(arr.get(0), |x| arr.get(**x as usize))
            .take_while(|x| **x != 0)
            .fold(0, |acc, x| 10 * acc + x + 1),
    )
}

pub fn part2<'a, I, S>(lines: I) -> Option<u64>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let nums = parse(lines)?;
    let mut arr = Vec::with_capacity(1000000);
    arr.extend(1..=1000000);
    for (i, x) in nums.iter().enumerate() {
        arr[x.checked_sub(1)? as usize] = *nums.get(i + 1).unwrap_or(&10) - 1;
    }
    let mut x = nums[0] - 1;
    *arr.last_mut()? = x;
    for _ in 0..10000000 {
        x = step::<1000000>(&mut arr[..], x)?;
    }
    Some((arr[0] as u64 + 1) * (arr[arr[0] as usize] as u64 + 1))
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn part1_examples() {
        assert_eq!(Some(67384529), part1(&["389125467"]));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(Some(149245887792), part2(&["389125467"]));
    }
}

File: py/src/aoc2020/day23.py
from array import array
import fileinput


def step(arr, x):
    a = arr[x]
    b = arr[a]
    c = arr[b]
    y = arr[c]
    t = x
    while True:
        t = t - 1 if t > 0 else len(arr) - 1
        if t not in (a, b, c):
            break
    u = arr[t]
    arr[x] = y
    arr[t] = a
    arr[c] = u
    return y


def part1(lines):
    '''
    >>> part1(('389125467', ))
    67384529
    '''
    arr = array('I', (0, ) * 9)
    nums = [int(c) for c in lines[0] if c.isdigit()]
    for i, x in enumerate(nums):
        arr[x - 1] = nums[(i + 1) % len(nums)] - 1
    x = nums[0] - 1
    for _ in range(100):
        x = step(arr, x)
    x, acc = 0, 0
    while True:
        x = arr[x]
        if not x:
            return acc
        acc = 10 * acc + x + 1


def part2(lines):
    '''
    >>> part2(('389125467', ))
    149245887792
    '''
    arr = array('I', range(1, 1000001))
    nums = [int(c) for c in lines[0] if c.isdigit()]
    for i, x in enumerate(nums):
        arr[x - 1] = nums[i + 1] - 1 if i + 1 < len(nums) else 9
    arr[-1] = nums[0] - 1
    x = nums[0] - 1
    for _ in range(10000000):
        x = step(arr, x)
    return (arr[0] + 1) * (arr[arr[0]] + 1)


parts = (part1, part2)

if __name__ == '__main__':
    lines = list(fileinput.input())
    print(part1(lines))
    print(part2(lines))
